### PR TITLE
[DROOLS-2766] support default values + allow to access evaluation con…

### PR DESCRIPTION
**Pull request by mariofusco**
…text in constraint evaluation + better error management in compiled DMN decision tables

### CODE REVIEW

**Comment by mariofusco** [comment]
Outstanding failures:

- null
DMNRuntimeTest.testLoan_Recommendation2
DMNRuntimeTest.testNull
DMNRuntimeTest.testNullOnNumber

- negate
DMNDecisionTableRuntimeTest.testDTUsingEqualsUnaryTestWithVariable1
DMNDecisionTableRuntimeTest.testDTUsingEqualsUnaryTestWithVariable2
DMNRuntimeTest.testListContainmentDT
DMNRuntimeTest.testNotListInDT

- question mark
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkAsString
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkInNonBooleanFunction
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkInstanceOf
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkMultivalue
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkMultivalueWithBrackets
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkOnly
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkVariableCorrectEvaluation
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkVariableVsQuestionMarkString
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithAnd
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithContext
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithForExpression
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithFuncDef
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithQuantifiedExpression:
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkInWithUnaryTests
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithNot
DMNDecisionTableRuntimeTest.testDecisionTablesQuestionMarkWithRange

- to be fixed
DMNDecisionTableRuntimeTest.testDecisionTableMultipleResults
DMNDecisionTableRuntimeTest.testNullRelation
DMNDecisionTableRuntimeTest.testSimpleDecisionTableMultipleOutputWrongOutputType
DMNRuntimeTest.testDrools2125
DMNRuntimeTest.testEventListeners
DMNRuntimeTest.testPriority_table
DMNRuntimeTest.testPriority_table_context_recursion
DMNRuntimeTest.testPriority_table_missing_output_values
DMNRuntimeTest.testWrongConstraintsInItemDefinition

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/DMNFEELHelper.java [hunk 1]
```diff
@@ -267,6 +268,10 @@ public Object evaluate(String expression, EvaluationContext ctx) {
         return feel.evaluate( expression, ctx );
     }
 
+    public Object evaluate(String expression, Map<String, Object> inputVariables) {
+        return feel.evaluate( expression, inputVariables );
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/DMNFEELHelper.java]: not necessary as part of this specific PR, but methods `feel.evaluate` accepting a String as an expression should not really be needed.

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/execmodelbased/ExecModelDMNEvaluatorCompiler.java [hunk 2]
```diff
@@ -202,7 +207,7 @@ public String generate( DMNCompilerContext ctx, DMNFEELHelper feel, DTableModel
                 for (int j = 0; j < dTableModel.getOutputSize(); j++) {
                     sb.append( "                            output" + j + ".add(evaluator.getOutput(" + i + ", " + j + "));\n" );
                 }
-                sb.append( "                            evaluator.registerFire(" + (i+1) + ");\n" );
+                sb.append( "                            evaluator.registerFire(" + i + ");\n" );
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/execmodelbased/ExecModelDMNEvaluatorCompiler.java:210]: why this went from `i+1` to `i` , please?

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/DMNInputRuntimeTest.java [hunk 3]
```diff
@@ -38,7 +32,18 @@
...
 
-public class DMNInputRuntimeTest {
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.kie.dmn.core.util.DynamicTypeUtils.entry;
+import static org.kie.dmn.core.util.DynamicTypeUtils.prototype;
+
+public class DMNInputRuntimeTest extends BaseDecisionTableTest {
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/DMNInputRuntimeTest.java]: nitpicking for the current stage, but the base class imho should need a better naming, something like `BaseInterpretedVsCompiledTest`

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/execmodelbased/ExecModelDMNEvaluatorCompiler.java (hunk 2 above)
Comment (mariofusco) [reply kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/execmodelbased/ExecModelDMNEvaluatorCompiler.java:210]: The HitPolicy class does that +1 internally. Also having a +1 there led me to an ArrayIndexOutOfBoundException.

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/DMNInputRuntimeTest.java (hunk 3 above)
Comment (mariofusco) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/DMNInputRuntimeTest.java]: Agreed, will rename the class with next PR.

File: kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/DMNFEELHelper.java [hunk 4]
```diff
@@ -279,7 +280,9 @@ public CompilerContext newCompilerContext() {
         return feel.newCompilerContext();
     }
 
-    public Object compile( String expr, CompilerContext feelctx ) {
-        return feel.compile( expr, feelctx );
+    public CompiledExpression compile( DMNModelImpl model, DecisionTable dtable, String dtableName, String expr, CompilerContext feelctx, int index ) {
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-core/src/main/java/org/kie/dmn/core/compiler/DMNFEELHelper.java:283]: Not necessary as part of this PR, but this method name is misleading. I think it should be called then "compileDTInputClause"

**Review by tarilabs — APPROVED** [review]

**Pull request merged**